messages/src/mgmt/mod.rs

```diff
@@ -0,0 +1,222 @@
+/// Accessing specific pages
+/// ------------------------
+///
+/// This network protocol requires that all memory accesses specify which page
+/// they would like to address. This is required so that the SP can correctly
+/// ask the actual free-side module to swap in the right page, and then read or
+/// write it as the host requested, all atomically from the point of view of the
+/// client(s).
+///
+/// > Important! It's the responsibility of the _host_ to determine if a given
+/// page or bank is supported by any particular module. We'd like to keep as
+/// much intelligence about parsing the memory maps in the host as possible,
+/// including identifying the management specification, and which pages if any
+/// are supported for a module. The intention is for the SP to only interpret
+/// sections of the map insofar as required for temperature and power
+/// monitoring.
+///
+/// Note that modules generally don't _fail_ a memory access to an unsupported
+/// page. Instead, modules generally just revert the page- or bank-select bytes
+/// to those known to be valid, usually zero. That means the client _must_ keep
+/// track of which bank and/or page is being accessed, and validate that those
+/// are actually supported by the module prior to operating on them.
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize, SerializedSize)]
+pub enum UpperPage {
+    Sff8636(sff8636::Page),
+    Cmis(cmis::Page),
+}
+
+impl UpperPage {
+    /// Return the actual page number of the upper page being mapped in.
+    pub fn page(&self) -> u8 {
+        match self {
+            UpperPage::Sff8636(inner) => inner.page(),
+            UpperPage::Cmis(inner) => inner.page(),
+        }
+    }
+
+    /// Return the bank of the upper page being accessed.
+    ///
+    /// Note that this returns `None` for pages that don't have a bank, e.g.,
+    /// those for modules conforming to SFF-8636.
+    ///
+    /// For those which _may_ have a bank, this _may_ return `Some(_)`. Note
+    /// that not all CMIS-defined pages allow a bank, so it may still return
+    /// `None` in that case.
+    pub fn bank(&self) -> Option<u8> {
+        match self {
+            UpperPage::Sff8636(_) => None,
+            UpperPage::Cmis(inner) => inner.bank(),
+        }
+    }
+}
+
+impl From<sff8636::Page> for UpperPage {
+    fn from(p: sff8636::Page) -> Self {
+        Self::Sff8636(p)
+    }
+}
+
+impl From<cmis::Page> for UpperPage {
+    fn from(p: cmis::Page) -> Self {
+        Self::Cmis(p)
+    }
+}
+
+/// A description of a region of a transceiver's memory map.
+///
+/// See [`UpperPage`] for a detailed description of how a module's upper page is
+/// mapped in for an operation. This type also performs validation against the
+/// offset / length, to ensure they're a valid access for the 256-byte memory
+/// map of all transceiver modules we support.
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize, SerializedSize)]
+pub struct MemoryRegion {
```

Comment:
General question: if the SP receives a `Read(MemoryRegion)` command over the network, is _it_ responsible for handling the memory region (e.g. configuring pages and banks)?

That's what's implied by our API, but I'm not sure whether the BF SDE expects that to be happening, or whether the SDE wants to manage pages itself (i.e. through manual reads and writes to `0x7E` and `0x7F`).

Reply:
That's correct. I should make that clearer in the message definitions.

It's really just about atomicity from the point of view of the host. If we make those separate requests, then theoretically we could interleave two separate requests that set the page, and neither would be accessing the correct data.

The question about SDE integration is a good one. It's not clear exactly how they use it, but they do expect to be able to write some register, potentially the page-select, and then read something else from that selected page. My plan to deal with that is to cache the requested page that the SDE thinks it's setting, and then always send it along in the request to the SP to ensure it's atomic.

Hopefully that clears things up, but I could very well be missing other considerations.

Reply:
This kind of stateful system is always tricky, but I think the plan sounds reasonable.  On the SP side, I'll look at caching configured page and banks, so that we don't do a bunch of extra transactions.

Reply:
I think by "extra transactions", do you mean to avoid an I2C write to set the page-select, if we're not changing the page from the previous read? I think that should be OK, but maybe unnecessary. I have basically zero idea what the frequency of these messages will be, so an argument could be made to defer that. Whatever you think is best!

Would more documentation be helpful, either on these variants or a crate comment?